node/compile-hook.js

```diff
@@ -3,78 +3,111 @@
 const exts = Module._extensions;
 const Mp = Module.prototype;
 
 let compileOptions;
+const reifyVersion = utils.getReifyVersion();
 
 module.exports = exports = (options) => {
   if (compileOptions === void 0) {
     compileOptions = Object.assign({}, options);
   }
 };
 
-function getCacheKey(filename) {
-  const mtime = utils.mtime(filename);
-  if (mtime < 0) {
-    return null;
+function addWrapper(func, wrapper) {
+  const wrappers = func.reified.wrappers;
+  if (typeof wrappers[reifyVersion] !== "function") {
+    wrappers[reifyVersion] = wrapper;
+  }
+}
+
+function createWrapperManager(object, property) {
+  const func = object[property];
+  if (! isManaged(func)) {
+    (object[property] = function(param, filename) {
+      const wrapper = getWrapper(object[property], filename);
+      return wrapper === null
+        ? func.call(this, param, filename)
+        : wrapper.call(this, func, param, filename);
+    }).reified = createWrapperMap(func);
   }
-  return {
-    filename,
-    mtime,
-    source: "Module.prototype._compile"
-  };
 }
 
-// Override Module.prototype._compile to compile any code that will be
-// evaluated as a module.
-const _compile = Mp._compile;
-if (! _compile.reified) {
-  (Mp._compile = function (content, filename) {
-    const options = {
-      filename,
-      cacheKey() {
-        return getCacheKey(filename);
-      }
-    };
-
-    content = compiler.compile(content, Object.assign(options, compileOptions));
-    return _compile.call(this, content, filename);
-  }).reified = _compile;
+function createWrapperMap(func) {
+  const map = new FastObject;
+  map.raw = func;
+  map.wrappers = new FastObject;
+  return map;
 }
 
-const extJs = exts[".js"];
-if (! (extJs && extJs.reified)) {
-  (exts[".js"] = (module, filename) => {
-    const pkgInfo = typeof filename === "string"
-      ? utils.getPkgInfo(path.dirname(filename))
-      : null;
-
-    const cachePath = pkgInfo === null ? null : pkgInfo.cachePath;
-    if (typeof cachePath === "string") {
-      const cache = pkgInfo.cache;
-      const cacheKey = getCacheKey(filename);
-      const cacheFilename = utils.getCacheFilename(cacheKey, pkgInfo.config);
-
-      let cacheValue = cache[cacheFilename];
-      if (cacheValue === true) {
-        const cacheFilepath = path.join(cachePath, cacheFilename);
-        const buffer = utils.readFile(cacheFilepath);
-        cacheValue = cache[cacheFilename] = utils.gunzip(buffer, "utf8");
-      }
-      if (typeof cacheValue === "string") {
-        return module._compile(cacheValue, filename);
-      }
+function getCacheKey(filename) {
+  const mtime = utils.mtime(filename);
+  return mtime < 0 ? null : { filename, mtime };
+}
+
+function getWrapper(func, filename) {
+  const pkgInfo = typeof filename === "string"
+    ? utils.getPkgInfo(path.dirname(filename))
+    : null;
+
+  if (pkgInfo !== null) {
+    const wrappers = func.reified.wrappers;
+    const versions = Object.keys(wrappers);
```

Comment:
If `Object.keys` is a perf concern we can maintain a separate array of versions. I haven't done a perf check of this stuff yet. But it's nothing that can't be addressed as needed.